File: 2017/day20.py
import math
from pathlib import Path
import re
import time


class Vec3:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z
        self.mag = math.sqrt(x ** 2 + y ** 2 + z ** 2)

    def __add__(self, other):
        return Vec3(self.x + other.x, self.y + other.y, self.z + other.z)

    def manhattan(self, other):
        return abs(self.x - other.x) + abs(self.y - other.y) + abs(self.z - other.z)


class Particle:
    def __init__(self, pos: Vec3, vel: Vec3, accel: Vec3, idx: int):
        self.pos = pos
        self.vel = vel
        self.accel = accel
        self.idx = idx

    def tick(self):
        self.vel += self.accel
        self.pos += self.vel


def main():
    particles = [parse_particle(line, i) for i, line in enumerate(
        Path("inputs/day20.txt").read_text().splitlines())]
    particles.sort(key=lambda x: (x.accel.mag, x.vel.mag, x.pos.mag))
    print([f"{p.idx}: {p.accel.mag}" for p in particles[:5]])
    # Part 1: tried all the lowest accel

    loop_start = time.time()
    dead = set()
    for _ in range(100):
        for i in range(len(particles) - 1):
            p1 = particles[i]
            if p1 in dead:
                continue
            p1_dead = False
            for j in range(i + 1, len(particles)):
                p2 = particles[j]
                if p2 in dead:
                    continue
                dist = p1.pos.manhattan(p2.pos)
                if dist == 0:
                    dead.add(p2)
                    p1_dead = True
            if p1_dead:
                dead.add(p1)
        for particle in particles:
            particle.tick()
    print(f"Part 2: {len(particles) - len(dead)}")
    loop_end = time.time()
    print(f"Loop time: {loop_end-loop_start:.4f} seconds")


def parse_particle(line: str, idx: int) -> Particle:
    sp = line.split(', ')
    pos_match = re.search(r"<.+>", sp[0])
    vel_match = re.search(r"<.+>", sp[1])
    accel_match = re.search(r"<.+>", sp[2])
    assert pos_match != None and vel_match != None and accel_match != None
    pos_nums = [int(num) for num in pos_match.group()[1:-1].split(',')]
    vel_nums = [int(num) for num in vel_match.group()[1:-1].split(',')]
    accel_nums = [int(num) for num in accel_match.group()[1:-1].split(',')]
    return Particle(
        Vec3(pos_nums[0], pos_nums[1], pos_nums[2]),
        Vec3(vel_nums[0], vel_nums[1], vel_nums[2]),
        Vec3(accel_nums[0], accel_nums[1], accel_nums[2]),
        idx
    )


if __name__ == "__main__":
    start = time.time()
    main()
    end = time.time()
    print(f"Execution time: {end-start:.4f} seconds")

File: 2017/day20.js
const fs = require("fs");

class Particle {
  constructor(pos, vel, acc, idx) {
    this.pos = pos;
    this.vel = vel;
    this.acc = acc;
    this.idx = idx;
  }

  tick() {
    this.vel = add(this.vel, this.acc);
    this.pos = add(this.pos, this.vel);
  }
}

console.time();

let particles = fs
  .readFileSync("inputs/day20.txt", "utf-8")
  .split("\n")
  .map((line, idx) => parseParticle(line, idx))
  .sort((a, b) => {
    let accMagA = mag(a.acc);
    let accMagB = mag(b.acc);
    if (accMagA !== accMagB) {
      return accMagA - accMagB;
    }
    let velMagA = mag(a.vel);
    let velMagB = mag(b.vel);
    return velMagA - velMagB;
  });
console.log(particles.slice(0, 5).map((p) => [p.idx, mag(p.acc)]));
// part 1, tried all lowest accel

let dead = new Set();
for (let iter = 0; iter < 100; iter++) {
  for (let i = 0; i < particles.length - 1; i++) {
    let p1 = particles[i];
    if (dead.has(p1)) continue;
    let p1Dead = false;
    for (let j = i + 1; j < particles.length; j++) {
      let p2 = particles[j];
      if (dead.has(p2)) continue;
      let dist = manhattan(p1.pos, p2.pos);
      if (dist === 0) {
        dead.add(p2);
        p1Dead = true;
      }
    }
    if (p1Dead) dead.add(p1);
  }
  for (let particle of particles) particle.tick();
}
console.log(`Part 2: ${particles.length - dead.size}`);

console.timeEnd();

function parseParticle(line, idx) {
  let groups = [...line.matchAll(/<([^>+]+)>/g)];
  let pos = groups[0][1].split(",").map((num) => parseInt(num));
  let vel = groups[1][1].split(",").map((num) => parseInt(num));
  let acc = groups[2][1].split(",").map((num) => parseInt(num));
  return new Particle(pos, vel, acc, idx);
}

function mag(vec) {
  return Math.sqrt(vec[0] ** 2 + vec[1] ** 2 + vec[2] ** 2);
}

function add(v1, v2) {
  return [v1[0] + v2[0], v1[1] + v2[1], v1[2] + v2[2]];
}

function manhattan(v1, v2) {
  return (
    Math.abs(v1[0] - v2[0]) + Math.abs(v1[1] - v2[1]) + Math.abs(v1[2] - v2[2])
  );
}
